Enable ending the task directly from the triggerer without going into the worker. (#40084)
mistercrunch/Airflow: docs/apache-airflow/authoring-and-scheduling/deferring.rst

            # We have no more work to do here. Mark as complete.
            return

Exiting deferred task from Triggers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 .. versionadded:: 2.10.0

> If you want to exit your task directly from the triggerer without going into the worker, you can specific operator level attribute ``end_from_trigger`` with the attributes to your deferrable operator other discussed above.

Reviewer: Is the name `end_from_trigger` a convention or a requirement?

Author: It's a convention.

Triggers can have two options: they can either send execution back to the worker or end the task instance directly. If the trigger ends the task instance itself, the ``method_name`` does not matter and can be ``None``. Otherwise, provide the name of the ``method_name`` that should be used when resuming execution in the task.

.. code-block:: python

    class DateTimeSensorAsync(DateTimeSensor):
        def __init__(self, end_from_trigger: bool = True, **kwargs) -> None:
            super().__init__(**kwargs)
            self.end_from_trigger = end_from_trigger

        def execute(self, context: Context) -> NoReturn:
            self.defer(

>                 method_name="execute_complete",

Reviewer: What if we only want to support the `end_from_tirgger` mode? Should we make it `None` or `__trigger_exit__`?

Author: It can be anything and it doesn't matter actually.

        trigger_kwargs=[{"moment": timedelta(hours=2)}, {"moment": timedelta(hours=2)}]
    )

Exiting deferred task from Triggers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 .. versionadded:: 2.10.0

> If you want to exit your task directly from the triggerer without going into the worker, you can specific operator level attribute ``end_from_trigger`` with the attributes to your deferrable operator other discussed above. This can save some resources needed to start a new worker.

Reviewer: ```suggestion
If you want to exit your task directly from the triggerer without going into the worker, you can specify the instance level attribute ``end_from_trigger`` with the attributes of your deferrable operator, as discussed above. This can save some resources needed to start a new worker.
```

I feel it's not actually operator-level. The example looks like an instance-level attribute

        trigger_kwargs=[{"moment": timedelta(hours=2)}, {"moment": timedelta(hours=2)}]
    )

Exiting deferred task from Triggers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 .. versionadded:: 2.10.0

> If you want to exit your task directly from the triggerer without going into the worker, you can specific operator level attribute ``end_from_trigger`` with the attributes to your deferrable operator other discussed above.

Reviewer 2: ```suggestion
If you want to exit your task directly from the triggerer without going into the worker, you can specific operator level attribute ``end_from_trigger`` with the attributes to your deferrable operator other discussed above. This can save some resources needed to start a new worker.
```

> Triggers can have two options: they can either send execution back to the worker or end the task instance directly. If the trigger ends the task instance itself, the ``method_name`` does not matter and can be ``None``. Otherwise, provide the name of the ``method_name`` that should be used when resuming execution in the task.

Reviewer 2: ```suggestion
Triggers can have two options: they can either send execution back to the worker or end the task instance directly. If the trigger ends the task instance itself, the ``method_name`` does not matter and can be ``None``. Otherwise, provide ``method_name`` that should be used when resuming execution in the task.
```

.. code-block:: python

>     class DateTimeSensorAsync(DateTimeSensor):
>         def __init__(self, end_from_trigger: bool = True, **kwargs) -> None:
>             super().__init__(**kwargs)
>             self.end_from_trigger = end_from_trigger
> 
>         def execute(self, context: Context) -> NoReturn:
>             self.defer(
>                 method_name="execute_complete",
>                 trigger=DateTimeTrigger(
>                     moment=timezone.parse(self.target_time), end_from_trigger=self.end_from_trigger
>                 ),
>             )
> 
>         def execute_complete(self, context, event=None) -> None:
>             return None

Reviewer 2: ```suggestion
    class DateTimeSensorAsync(DateTimeSensor):
        def __init__(self, **kwargs) -> None:
            super().__init__(**kwargs)
            self.end_from_trigger = True

        def execute(self, context: Context) -> NoReturn:
            self.defer(
                method_name=None,
                trigger=DateTimeTrigger(
                    moment=timezone.parse(self.target_time), end_from_trigger=self.end_from_trigger
                ),
            )
```

Let’s make this example simple for people. If users want their sensor to support both modes, I think they should be smart enough to add the argument.

(We should also add text to tell users this sensor _always_ exit from trigger.)

Author: I have changed the example completely now with `WaitFiveHourTrigger` and `WaitFiveHourSensorAsync`.

``TaskSuccessEvent`` and ``TaskFailureEvent`` are the two events that can be used to end the task instance directly. This marks the task with the state ``task_instance_state`` and optionally pushes xcom if applicable. Here's an example of how to use these events:

.. code-block:: python

    class DateTimeTrigger(BaseTrigger):

>         def __init__(self, moment: datetime.datetime, *, end_from_trigger: bool = False):

Reviewer 2: Same here, we can make the example simpler so it is easier to understand.

Reviewer 2: Maybe we should have a better example than datatime in the first place; people don’t ever need to implement a datetime sensor because Airflow already has one.

Author: I have changed the example completely now with `WaitFiveHourTrigger` and `WaitFiveHourSensorAsync`.
